### PR TITLE
Fix using client configuration flags with in-cluster configuration 

### LINKED ISSUE #93474: kubectl --request-timeout (or any other client configuration flag) prevents use of in-cluster configuration
**What happened**:
When running a pod within the cluster, using the --request-timeout option always fails instantly

**What you expected to happen**:
It should respect --request-timeout (or, at a minimum, give a better error message).  The real issue is that one can not make sure that kubectl will timeout.

We expect to need to timeout under adverse conditions but kubectl in-cluster never times out and thus we have to do process-level timeouts and kill operations.

**How to reproduce it (as minimally and precisely as possible)**:
Create a trivial container that can run kubectl and give the pod a service role to access "kubectl get pods" (for the example below).  It does not matter if you give full access or limited access roles - the --request-timeout option always causes kubectl to fail when run within the cluster.

**Anything else we need to know?**:

**Environment**:
- Kubernetes version (use `kubectl version`):
  Client Version: version.Info{Major:"1", Minor:"18", GitVersion:"v1.18.4", GitCommit:"c96aede7b5205121079932896c4ad89bb93260af", GitTreeState:"clean", BuildDate:"2020-06-17T15:57:19Z", GoVersion:"go1.13.6", Compiler:"gc", Platform:"linux/amd64"}
  Server Version: version.Info{Major:"1", Minor:"18", GitVersion:"v1.18.4", GitCommit:"c96aede7b5205121079932896c4ad89bb93260af", GitTreeState:"clean", BuildDate:"2020-06-17T15:57:19Z", GoVersion:"go1.13.6", Compiler:"gc", Platform:"linux/amd64"}

- Cloud provider or hardware configuration: Azure
- OS (e.g: `cat /etc/os-release`): ubuntu 18.04.4 LTS
- Kernel (e.g. `uname -a`): Linux k8s-master-18861755-2 5.3.0-1031-azure #32~18.04.1-Ubuntu SMP Mon Jun 22 15:27:23 UTC 2020 x86_64 x86_64 x86_64 GNU/Linux
- Install tools: aks-engine 0.53.0
- Network plugin and version (if this is a network-related bug): Azure CNI
- Others:

In a cluster, on a pod, I can issue `kubectl get pods` without trouble but if it includes `--request-timeout` it instantly fails:

```
# kubectl version
Client Version: version.Info{Major:"1", Minor:"18", GitVersion:"v1.18.4", GitCommit:"c96aede7b5205121079932896c4ad89bb93260af", GitTreeState:"clean", BuildDate:"2020-06-17T15:57:19Z", GoVersion:"go1.13.6", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"18", GitVersion:"v1.18.4", GitCommit:"c96aede7b5205121079932896c4ad89bb93260af", GitTreeState:"clean", BuildDate:"2020-06-17T15:57:19Z", GoVersion:"go1.13.6", Compiler:"gc", Platform:"linux/amd64"}
# kubectl get pods --request-timeout 30
The connection to the server localhost:8080 was refused - did you specify the right host or port?
# kubectl get pods
NAME                                                              READY   STATUS      RESTARTS   AGE
acr-transformer-6f497d647c-5rx7d                                  1/1     Running     0          18m
acr-transformer-6f497d647c-d9j2s                                  1/1     Running     0          18m
. . . .
```

### PR DESCRIPTION
**Pull request by npolshakova**
#### What type of PR is this?

/kind bug

#### What this PR does / why we need it:

Currently kubectl --request-timeout (or any other client configuration flag) does not work with in-cluster configuration because isDefaultConfig returns false when [checking for equality](https://github.com/kubernetes/kubernetes/blob/63eb5028ba613003f044f0391912be356e0f1b97/staging/src/k8s.io/client-go/tools/clientcmd/loader.go#L382) and then the [mergedConfig](https://github.com/kubernetes/kubernetes/blob/63eb5028ba613003f044f0391912be356e0f1b97/staging/src/k8s.io/client-go/tools/clientcmd/merged_client_builder.go#L115) is used instead of the [in cluster config](https://github.com/kubernetes/kubernetes/blob/63eb5028ba613003f044f0391912be356e0f1b97/staging/src/k8s.io/client-go/tools/clientcmd/merged_client_builder.go#L120). 

#### Which issue(s) this PR fixes:

Fixes https://github.com/kubernetes/kubernetes/issues/93474 

#### Special notes for your reviewer:

You can reproduce this bug in [kind](https://kind.sigs.k8s.io/) with a pod container with kubectl such as `bitnami/kubectl`. I validated the fix by building a docker container with kubectl:

1. Cross compile kubectl: 

```
build/run.sh make kubectl KUBE_BUILD_PLATFORMS=linux/amd64
```

2. Build container image with new kubectl: 

```
docker build -t nina-kubectl -f test.Dockerfile .
```

The dockerfile I used:
```
FROM ubuntu:20.04

WORKDIR /app

COPY kubectl /usr/local/bin/
RUN chmod +x /usr/local/bin/kubectl

USER 1001
ENTRYPOINT [ "kubectl" ]
CMD [ "--help" ]
``` 

3. Run a local kind cluster and load the image  with `kind load docker-image nina-kubectl:latest` 
4. Create a deployment, and correct rbac permissions to run `kubectl get pods`.

```
apiVersion: apps/v1
kind: Deployment
metadata:
  name: kubectl-deployment
spec:
  replicas: 1
  selector:
    matchLabels:
      app: kubectl
  template:
    metadata:
      labels:
        app: kubectl
    spec:
      serviceAccountName: my-service-account
      containers:
      - name: kubectl-container
        image: nina-kubectl:latest
        command: [ "/bin/sh", "-c", "--" ]
        args: [ "while true; do sleep 30; done;" ]
```

5. Exec into pod and test kubectl behavior with and without client config with fix: 

```
I have no name!@kubectl-deployment-6874f6db87-9s9qb:/app$ kubectl get pods
NAME                                    READY   STATUS        RESTARTS   AGE
kubectl-deployment-6874f6db87-9s9qb     1/1     Running       0          15s
kubectl-deployment-6874f6db87-mfskr     1/1     Terminating   0          61s
pod-identity-webhook-8565f74769-2772d   1/1     Running       0          27h
I have no name!@kubectl-deployment-6874f6db87-9s9qb:/app$ kubectl get pods  --request-timeout 30
NAME                                    READY   STATUS    RESTARTS   AGE
kubectl-deployment-6874f6db87-9s9qb     1/1     Running   0          35s
pod-identity-webhook-8565f74769-2772d   1/1     Running   0          27h
```

### New kubectl e2e test cases

It was a little tricky running the kubectl e2e tests. It looks like the in-cluster tests only run on amd64 setups and require the steps to run locally: 
1. Make sure you have a kind cluster (`kind create cluster`)
2. Build the e2e.test (`make all WHAT=test/e2e/e2e.test vendor/github.com/onsi/ginkgo/v2/ginkgo`)
3. Build `kubectl` and set it on your PATH (otherwise it used the default system kubectl without the fix)
4. Focus the test and run with: `KUBERNETES_PROVIDER=skeleton ./hack/ginkgo-e2e.sh --ginkgo.focus="should handle in-cluster config" --ginkgo.skip='\[Serial\]|\[Deprecated\]'`

The new e2e test does a similar check by setting the `--request-timeout`. If testing locally you don't set the new kubectl in the path, it will use the `/usr/local/bin/kubectl` kubectl and fail on the old behavior: 

```
  [FAILED] error running /usr/local/bin/kubectl --server=https://127.0.0.1:38847 --kubeconfig=/home/npolshak/.kube/config --namespace=kubectl-8070 exec httpd -- /bin/sh -x -c /tmp/kubectl get pods --request-timeout=10s --v=6 2>&1:
  Command stdout:
  I0226 21:47:29.010979     177 merged_client_builder.go:163] Using in-cluster namespace
  I0226 21:47:29.011688     177 round_trippers.go:553] GET http://localhost:8080/api?timeout=10s  in 0 milliseconds
  E0226 21:47:29.011738     177 memcache.go:265] couldn't get current server API group list: Get "http://localhost:8080/api?timeout=10s": dial tcp [::1]:8080: connect: connection refused
  I0226 21:47:29.011747     177 cached_discovery.go:120] skipped caching discovery info due to Get "http://localhost:8080/api?timeout=10s": dial tcp [::1]:8080: connect: connection refused
  I0226 21:47:29.012001     177 round_trippers.go:553] GET http://localhost:8080/api?timeout=10s  in 0 milliseconds
  E0226 21:47:29.012034     177 memcache.go:265] couldn't get current server API group list: Get "http://localhost:8080/api?timeout=10s": dial tcp [::1]:8080: connect: connection refused
  I0226 21:47:29.013145     177 cached_discovery.go:120] skipped caching discovery info due to Get "http://localhost:8080/api?timeout=10s": dial tcp [::1]:8080: connect: connection refused
  I0226 21:47:29.013163     177 shortcut.go:100] Error loading discovery information: Get "http://localhost:8080/api?timeout=10s": dial tcp [::1]:8080: connect: connection refused
  I0226 21:47:29.013583     177 round_trippers.go:553] GET http://localhost:8080/api?timeout=10s  in 0 milliseconds
  E0226 21:47:29.013625     177 memcache.go:265] couldn't get current server API group list: Get "http://localhost:8080/api?timeout=10s": dial tcp [::1]:8080: connect: connection refused
  I0226 21:47:29.014716     177 cached_discovery.go:120] skipped caching discovery info due to Get "http://localhost:8080/api?timeout=10s": dial tcp [::1]:8080: connect: connection refused
  I0226 21:47:29.015128     177 round_trippers.go:553] GET http://localhost:8080/api?timeout=10s  in 0 milliseconds
  E0226 21:47:29.015167     177 memcache.go:265] couldn't get current server API group list: Get "http://localhost:8080/api?timeout=10s": dial tcp [::1]:8080: connect: connection refused
  I0226 21:47:29.016273     177 cached_discovery.go:120] skipped caching discovery info due to Get "http://localhost:8080/api?timeout=10s": dial tcp [::1]:8080: connect: connection refused
  I0226 21:47:29.016651     177 round_trippers.go:553] GET http://localhost:8080/api?timeout=10s  in 0 milliseconds
  E0226 21:47:29.016696     177 memcache.go:265] couldn't get current server API group list: Get "http://localhost:8080/api?timeout=10s": dial tcp [::1]:8080: connect: connection refused
  I0226 21:47:29.017882     177 cached_discovery.go:120] skipped caching discovery info due to Get "http://localhost:8080/api?timeout=10s": dial tcp [::1]:8080: connect: connection refused
  I0226 21:47:29.017915     177 helpers.go:264] Connection error: Get http://localhost:8080/api?timeout=10s: dial tcp [::1]:8080: connect: connection refused
  The connection to the server localhost:8080 was refused - did you specify the right host or port?

  stderr:
  + /tmp/kubectl get pods '--request-timeout=10s' '--v=6'
  command terminated with exit code 1
```

If you set kubectl to point to the locally built one (ie. `_output/dockerized/bin/linux/amd64`) the test passes as expected:

```
  STEP: getting pods with in-cluster configs and request timeout set @ 02/26/24 16:37:54.548
  I0226 16:37:54.548134 529679 builder.go:121] Running '/home/npolshak/kubernetes/_output/dockerized/bin/linux/amd64/kubectl --server=https://127.0.0.1:38847 --kubeconfig=/home/npolshak/.kube/config --namespace=kubectl-8167 exec httpd -- /bin/sh -x -c /tmp/kubectl get pods --request-timeout=10s --v=6 2>&1'
  I0226 16:37:54.720737 529679 builder.go:146] stderr: "+ /tmp/kubectl get pods '--request-timeout=10s' '--v=6'\n"
  I0226 16:37:54.720795 529679 builder.go:147] stdout: "I0226 21:37:54.705002     185 merged_client_builder.go:184] Using in-cluster namespace\nI0226 21:37:54.705260     185 merged_client_builder.go:133] Using in-cluster configuration\nI0226 21:37:54.707998     185 merged_client_builder.go:133] Using in-cluster configuration\nI0226 21:37:54.713937     185 round_trippers.go:553] GET https://10.96.0.1:443/api/v1/namespaces/kubectl-8167/pods?limit=500&timeout=10s 200 OK in 5 milliseconds\nNAME    READY   STATUS    RESTARTS   AGE\nhttpd   1/1     Running   0          25s\n"
  I0226 16:37:54.720810 529679 output.go:94] stdout: I0226 21:37:54.705002     185 merged_client_builder.go:184] Using in-cluster namespace
  I0226 21:37:54.705260     185 merged_client_builder.go:133] Using in-cluster configuration
  I0226 21:37:54.707998     185 merged_client_builder.go:133] Using in-cluster configuration
  I0226 21:37:54.713937     185 round_trippers.go:553] GET https://10.96.0.1:443/api/v1/namespaces/kubectl-8167/pods?limit=500&timeout=10s 200 OK in 5 milliseconds
  NAME    READY   STATUS    RESTARTS   AGE
  httpd   1/1     Running   0          25s
``` 

The second new e2e test checks that providing an invalid user with `--as=fake` with in-cluster config fails as expected with a 403 error: 
```
  STEP: trying to use kubectl with fake user @ 02/26/24 17:03:20.469
  I0226 17:03:20.469059 534536 builder.go:121] Running '/home/npolshak/kubernetes/_output/dockerized/bin/linux/amd64/kubectl --server=https://127.0.0.1:38847 --kubeconfig=/home/npolshak/.kube/config --namespace=kubectl-2411 exec httpd -- /bin/sh -x -c /tmp/kubectl get pods --as=fake --v=6 2>&1'
  I0226 17:03:20.644976 534536 builder.go:135] rc: 1
  I0226 17:03:20.645028 534536 kubectl.go:657] got err error running /home/npolshak/kubernetes/_output/dockerized/bin/linux/amd64/kubectl --server=https://127.0.0.1:38847 --kubeconfig=/home/npolshak/.kube/config --namespace=kubectl-2411 exec httpd -- /bin/sh -x -c /tmp/kubectl get pods --as=fake --v=6 2>&1:
  Command stdout:
  I0226 22:03:20.629839     184 merged_client_builder.go:184] Using in-cluster namespace
  I0226 22:03:20.630053     184 merged_client_builder.go:133] Using in-cluster configuration
  I0226 22:03:20.633811     184 merged_client_builder.go:133] Using in-cluster configuration
  I0226 22:03:20.637789     184 round_trippers.go:553] GET https://10.96.0.1:443/api/v1/namespaces/kubectl-2411/pods?limit=500 403 Forbidden in 3 milliseconds
  I0226 22:03:20.638087     184 helpers.go:246] server response object: [{
    "kind": "Status",
    "apiVersion": "v1",
    "metadata": {},
    "status": "Failure",
    "message": "users \"fake\" is forbidden: User \"system:serviceaccount:kubectl-2411:default\" cannot impersonate resource \"users\" in API group \"\" at the cluster scope",
    "reason": "Forbidden",
    "details": {
      "name": "fake",
      "kind": "users"
    },
    "code": 403
  }]
  Error from server (Forbidden): users "fake" is forbidden: User "system:serviceaccount:kubectl-2411:default" cannot impersonate resource "users" in API group "" at the cluster scope

  stderr:
  + /tmp/kubectl get pods '--as=fake' '--v=6'
  command terminated with exit code 1

  error:
  exit status 1
```

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

### CODE REVIEW

**Comment by alexzielenski** [comment]
/assign @seans3 
could you check this out if you have time?

/triange accepted

**Comment by knelasevero** [comment]
/retest

I was working on a similar fix but forgot to send a comment in that issue :D 

I was going to push now and just saw your PR, thanks!

I will check if my test cases have anything different, but skimming through looks very similar to yours

File: staging/src/k8s.io/client-go/tools/clientcmd/loader.go [hunk 1]
```diff
@@ -68,7 +68,7 @@ func currentMigrationRules() map[string]string {
 type ClientConfigLoader interface {
 	ConfigAccess
 	// IsDefaultConfig returns true if the returned config matches the defaults.
-	IsDefaultConfig(*restclient.Config) bool
+	IsDefaultConfig(*restclient.Config) (bool, error)
```
Comment (aojea) [inline staging/src/k8s.io/client-go/tools/clientcmd/loader.go:71]: is this really needed?
it will be nice to not have to modify a public signature

Comment (npolshakova) [reply staging/src/k8s.io/client-go/tools/clientcmd/loader.go:71]: It's not really needed to fix the bug and I can revert it, but then the unit tests will need to be changed for the `in-cluster not checked on non-empty error` test case. 

In the current implementation if the config is invalid, I'm skipping the in cluster check [here](https://github.com/kubernetes/kubernetes/blob/40e35e611983cea2fa7b150871be355fb5558063/staging/src/k8s.io/client-go/tools/clientcmd/merged_client_builder.go#L116). So keeping the signature the same would add a little additional overhead for checking the in-cluster case if the config is invalid.

Comment (aojea) [reply staging/src/k8s.io/client-go/tools/clientcmd/loader.go:71]: well, it does not seem widely used https://grep.app/search?current=3&q=IsDefaultConfig&filter[lang][0]=Go

so may be is not a big problem

**Pull request closed without merging**